src/Interpreters/UserDefinedFunctionsVisitor.cpp

```diff
@@ -0,0 +1,99 @@
+#include "UserDefinedFunctionsVisitor.h"
+
+#include <unordered_map>
+#include <stack>
+
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTCreateFunctionQuery.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/UserDefinedFunctionFactory.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNSUPPORTED_METHOD;
+}
+
+void UserDefinedFunctionsMatcher::visit(ASTPtr & ast, Data &)
+{
+    auto * function = ast->as<ASTFunction>();
+    if (!function)
+        return;
+
+    auto result = tryToReplaceFunction(*function);
+    if (result)
+        ast = result;
+}
+
+bool UserDefinedFunctionsMatcher::needChildVisit(const ASTPtr &, const ASTPtr &)
+{
+    return true;
+}
+
+ASTPtr UserDefinedFunctionsMatcher::tryToReplaceFunction(const ASTFunction & function)
```

Comment:
So it's still an AST-based solution. I think we can build a real function using ActionsDAG for lambdas.

Reply:
@amosbird it will be weaker solution, because for example client created function.
```
CREATE FUNCTION test_func (a, b) -> a + b;
```
It is used like
```
SELECT test_func(a, b) + c;
```
In our solution it will be substituted and whole expression can be optimised later for example JIT compiled. There are much more cases when we get benefits from substitution, this one is the simples.

Reply:
Great. I'm also a fan of AST based solutions.